### PR TITLE
Improve error message with mismatched valtype/dtype

### LINKED ISSUE #2670: Expected ... to be bound to an at::Tensor, but got l
```
Traceback (most recent call last):
  File "/home/tfogal/env/lib/python3.10/site-packages/nvfuser/__init__.py", line 145, in execute
    result = self._execute(
RuntimeError: Expected T0_g[ iS0{i1}, iS1{i2}, iS2{i3}, iS3{i4} ] to be bound to an at::Tensor, but got l
Exception raised from validateValWithConcreteValue at /tmp/pip-req-build-fk_mwug5/csrc/expr_evaluator.cpp:29 (most recent call first):
frame #0: nvfuser::nvfCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xf3 (0x74a9a5f92f9b in /home/tfogal/env/lib/python3.10/site-packages/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #1: nvfuser::ExpressionEvaluator::bind_(nvfuser::Val const*, dynamic_type::DynamicType<dynamic_type::Containers<std::vector>, nvfuser::StructHandle, nvfuser::Pointer, nvfuser::Opaque, at::Tensor, std::complex<double>, double, long, bool>, bool) + 0x188a (0x74a9a632786a in /home/tfogal/env/lib/python3.10/site-packages/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #2: nvfuser::executor_utils::bindInputs(nvfuser::KernelArgumentHolder const&, nvfuser::Fusion*) + 0x287 (0x74a9a6316da7 in /home/tfogal/env/lib/python3.10/site-packages/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #3: <unknown function> + 0x69dee4 (0x74a9a6503ee4 in /home/tfogal/env/lib/python3.10/site-packages/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #4: nvfuser::FusionExecutorCache::runFusionWithInputs(c10::ArrayRef<c10::IValue> const&, std::optional<nvfuser::PrimDataType>, std::optional<signed char>) + 0x1e7 (0x74a9a6505827 in /home/tfogal/env/lib/python3.10/site-packages/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
```

Simplified reproducer:
```python
import torch
from nvfuser import FusionDefinition, DataType

def a(FusionDefinition):
    b.define_tensor([])
with FusionDefinition() as b:
    a(b)
c = [0]
b.execute(c)
```

[The original, larger program that failed](https://github.com/user-attachments/files/16365137/failure.py.txt)

### PR DESCRIPTION
**Pull request by jacobhinkle**
We already check types when binding arguments, but now we explicitly state which arguments mismatch, and in which positions.

Fixes #2670 by improving the error message. Before this change the error message was
```
Expected T0_g[ iS0{i1} ] to be bound to an at::Tensor, but got l
```
After this PR, the error is
```
Expected input 0, T0_g[ iS0{i0} ], to be an at::Tensor but got scalar 2
```

### CODE REVIEW

**Comment by jacobhinkle** [comment]
!build

**Review by tfogal — APPROVED** [review]
That error message is a great improvement! Thank you.

**Comment by jacobhinkle** [comment]
This is not an ideal fix since it replicates (poorly) the validation that already exists in `ExpressionEvaluator::bind`. I did it this way to be able to provide the argument number, which we lose once we get to the `bind` call. I'll try again to just improve the error message locally within `bind` and omit the argument number for now. In the future, it might be nice to be able to declare error contexts with guards so that additional info is printed. For example in `bindInputs` we could have:
```c++
for (const auto i : c10::irange(inputs.size())) {
  ErrorContext e("Error binding input " + std::to_string(i));
  expr_eval.bind(inputs[i], *args[i], true);
}
```
Then if `NVF_CHECK` or `NVF_ERROR` fails inside the lifetime of `e`, we would get an error message like `Error binding input 2: <actual error>`. That's a more general problem that we should handle separately (I'll post an issue for it). For now I will just try and clean up the `bind` error messages and revert the current changes.

**Comment by jacobhinkle** [comment]
I fixed this up by determining the input position inside the `bind_` call when a check fails. I also added some cases to the test to check that our error messages work properly for other kinds of malformed inputs.

File: csrc/expr_evaluator.cpp [hunk 1]
```diff
@@ -14,32 +14,60 @@
 #include <ir/iostream.h>
 #include <ir/utils.h>
 #include <logical_domain_map.h>
+#include <polymorphic_value.h>
 
+#include <cxxabi.h>
```
Comment (tfogal) [inline csrc/expr_evaluator.cpp:19]: Is this header file used / needed? maybe it was initially there for debugging?

**Review by tfogal — APPROVED** [review]
Re-+1'ing the new changes.

But I also want to plug #1758: I'd really encourage an up-front pass to validate as soon as we can. As we found with latency work (#2136), ExprEvaluator is on the critical path, and so I worry about adding more checks in there.

Admittedly--these checks *have* to be on the critical path (since we can only validate when we see what inputs we actually receive). But it would _probably_ behoove us to do an up-front check for UX purposes and avoid checks inside each Expr evaluation.

**Pull request merged**